README.md
# Working-with-data-R


history.txt
1505143926306:1
1505144032109:x = 1
1505144035963:del(x)
1505144228383:library("lattice", lib.loc="/Library/Frameworks/R.framework/Versions/3.4/Resources/library")
1505144230707:detach("package:lattice", unload=TRUE)
1505144240185:cls
1505144252096:c([1, 2, 3])
1505144256168:c(1, 2, 3)
1505144265399:x = c(1, 2, 3)
1505144267836:x = 1
1505144352262:getwd()
1505144354873:ls
1505144366956:rm
1505144370082:ls
1505144422857:getwd()
1505144437925:setwd('R Projects')
1505144440966:getwd()
1505144655175:?library
1505300486063:1
1505300495400:# Hi
1505300497691:Hi
1505300500522:#Hi
1505300504079:# Hi
1505300603804:x <- rnorm(100)
1505300604904:x
1505300608546:head(x)
1505300614656:mean(x)
1505300623515:sd(x)
1505300628810:min(x)
1505300631358:max(x)
1505300680339:tail(x)
1505649433990:library(foreign)
1505649440698:library(foreigns)
1505649500845:saldata <- read.spss("")
1505649504890:getwd()
1505649596276:saldata <- read.spss("/Users/jayadeepsasikumar/MSc in Data Analytics/Probability and Statistical Inference/Week 2 - Sep 18 to Sep 22/salarynorm.sav", use.value.labels=TRUE, max.value.labels=Inf, to.data.frame=TRUE)
1505649600805:saldata
1505649632988:colnames(saldata) <- tolower(colnames(saldata))
1505649643140:colnames(saldata)
1505649652990:colnames(saldata) <- toupper(colnames(saldata))
1505649654104:colnames(saldata)
1505649656624:colnames(saldata) <- tolower(colnames(saldata))
1505649657750:colnames(saldata)
1505649671002:saldata$gender
1505649687445:str(saldata$gender)
1505649700129:summary(saldata$gender)
1505649710368:summary(saldata$salary)
1505649731733:require(RcmdrMisc)
1505649768499:install.packages("Rcmdr")
1505728374382:1
1505728394989:x <- 1
1505728414017:y <- data.frame()
1505728516549:B = data.frame(x1=runif(100000), x2=runif(100000), x3=runif(100000))
1505728525658:summary(B)
1505728778815:library(help='base')
1505728834147:install.packages('devtools')
1505728894670:library(help='devtools')
1505729019947:install.packages('tinyverse')
1505729070083:install.packages('tidyverse')
1505729173957:x = 10
1505729174823:x
1505729180579:x <<- 12
1505729181164:x
1505729218750:x <- 1:10
1505729219550:x
1505729231519:typeof(x)
1505729236016:x ^ 2
1505744324685:x = 1
1505744327819:typeof(x)
1505744331076:x = 1L
1505744331914:typeof(x)
1505744367740:(0.5 - 0.4) == 0.1
1505744371002:(0.5 - 0.4)
1505744376302:x <- (0.5 - 0.4)
1505744380317:typeof(x)
1505744391707:0.1 == 0.1
1505744401356:typeof(0.1)
1505744437328:all.equal(0.5 - 0.4, 0.1)
1505744574969:x <- 'abcdefg'
1505744580507:x[1]
1505744583217:x[2]
1505744585572:x[3]
1505744587492:x[0]
1505744593353:typeof(x)
1505744772301:x
1505744777924:substr(x)
1505744787021:?substr
1505744800541:substr(x, 1, 7)
1505744802774:substr(x, 1, 8)
1505744804626:substr(x, 1, 6)
1505744808896:substr(x, 0, 7)
1505744811679:substr(x, -1, 7)
1505744818693:substr(x, 1, 2)
1505744827342:substr(x, 2, 6)
1505744848095:as.double(x)
1505744852539:y = as.double(x)
1505744854255:y
1505744887529:x = as.character('')
1505744899930:as.integer('x')
1505744905962:x <-as.integer('x')
1505744987806:as.double('1')
1505744997540:as.integer('1')
1505745000806:as.integer('1.2')
1505745003952:as.integer('1.8')
1505745017580:if
1505745023512:()
1505745028919:if T:
1505745030989:if T
1505745036055:?if
1505745039573:endif
1505745073445:if(T) x <- 1 else x <- 2
1505745077857:if(T) x <- 3 else x <- 2
1505745215746:y <- 1:10
1505745219447:typeof(y)
1505745226935:y <- 1.0:10.0
1505745227955:y
1505745230488:typeof
1505745235320:typeof(y)
1505745247542:vect
1505745260390:as.vector(y)
1505745264981:v = as.vector(y)
1505745265370:v
1505745269449:typeof(v)
1505745330393:x <- c(1, 2, 3, 4)
1505745331139:x
1505745333358:typeof(x)
1505745344841:as.integer(x)
1505745350141:x <- as.integer(x)
1505745350895:x
1505745352800:typeof
1505745354414:typeof(x)
1505745377373:x <- c(1L, 2, 3, 4)
1505745379601:typeof(x)
1505745385910:x <- c(1, 2, 3, 4L)
1505745387207:typeof(x)
1505745423605:x <- c('1', 2, 3, 4)
1505745424174:x
1505745426711:typeof(x)
1505745434478:x <- c(1, 2, 3, '4L')
1505745435029:x
1505745517213:x <- c(1, T)
1505745517675:x
1505745524622:typeof(x)
1505745532157:x <- c(1L, T)
1505745532532:x
1505745534323:typeof(x)
1505745541299:x <- c('1L', T)
1505745542314:x
1505745550284:as.character(T)
1505745565034:c(1, 2, 3, 4, T)
1505745568784:c(1, 2, 3, 4, F)
1505745612901:c(1.2, 1)
1505745614549:c(1.2, 1L)
1505745636972:x <- c(1.2, 1L)
1505745637773:x
1505745641035:typeof(x)
1505745936924:age = c(71, 56, 71, 71)
1505745953684:name = c('Donald', 'Barack', 'George', 'Bill')
1505745970591:democrat = c(F, T, F, T,)
1505745972184:democrat = c(F, T, F, T)
1505746049047:y = c(a=1, b=2, c=3, d=4)
1505746049908:y
1505746059164:y = c('a'=1, b=2, c=3, d=4)
1505746060108:y
1505746065311:y = c('a'=1, b=2, c=3, d=4, 5=5)
1505746074689:y = c('a'=1, b=2, c=3, d=4, 5)
1505746075828:y
1505746080803:y = c('a'=1, b=2, c=3, d=4, 5, f=6)
1505746082080:y
1505746094081:names(y)
1505746097447:names(age)
1505746106578:names(age) <- name
1505746108840:age
1505746118470:names(age) <- NULL
1505746120608:age
1505746123068:names(age) <- NULL
1505746128682:names <- name
1505746130686:names
1505746145053:names(age) = names
1505746148863:names
1505746152972:age
1505746185872:remove(names)
1505746189055:age
1505746201545:typeof(NULL)
1505746204672:NULL
1505746262179:length(age)
1505746273416:nchar(age)
1505746335910:?list
1505746368639:trump = c("Donald", 71, F)
1505746370146:trump
1505746377253:trump = list(trump)
1505746378740:trump
1505746397248:trump = list("Donald", 71, F)
1505746399404:trump
1505746422441:trump[[0]]
1505746427044:trump[[1]]
1505746431252:trump[[2]]
1505746434504:trump[[3]]
1505746526664:names(trump) <- c('name', 'age', 'democrat')
1505746530180:trump
1505746552057:names(trump) <- NULL
1505746562923:trump = list("Donald", 71, F)
1505746585995:?names
1505746619568:length(trumo)
1505746621348:length(trump)
1505746689617:x = c(1, 2, 3)
1505746702636:names(x) = c('a', 'b', 'c')
1505746703422:x

Plots.R
library(ggplot2)
library(readr)

PlotEmploymentTrends <- function (employment.trends) {
  # Plots the employment trends as per Q1.1
  qplot(x=time, y=employment, data=employment.trends, geom="line",
        xlab="Quarterly Figures", ylab="Trend",
        main="Dublin Employment Trends Per Sector: 2006-2016") +
    facet_grid(. ~ sector) +
    theme(axis.title.y = element_text(angle=0, vjust = 0.5))
}

PlotPropertyTrends <- function (property.trends) {
  # Plots the property trends as per Q1.2
  qplot(x=time, y=trend, data=property.trends, geom="line",
        colour=category, xlab="Time", ylab="Trend",
        main="Dublin Property Trends: 2007 - 2016") +
    labs(colour = 'Category') +
    theme(axis.title.y = element_text(angle=0, vjust = 0.5))
}

employment.trends <- read_delim("dublin employment trends.txt", delim=':')
colnames(employment.trends) <- tolower(colnames(employment.trends))

property.trends <- read_tsv("dublin property trends.txt")
colnames(property.trends) <- tolower(colnames(property.trends))


BikeInfo.R
library(dplyr)
library(ggmap)
library(ggplot2)
library(httr)
library(jsonlite)

kJCDAPIKey <- "f3af0867d0ae0de6957fa29b4469667fba4e229e"

ClassifyStation <- function (bike.stands) {
  # Classifies the bike station depending on the number of bike stands in it.
  #
  # Args:
  #   bike.stands: the number of bike stands in the station.
  #
  # Returns:
  #   Class of the station based on the number of bike stands - three possible
  #   classes are Small, Medium and Large.
  if (bike.stands <= 20) {
    return ("Small")
  } else if (bike.stands <= 30) {
    return ("Medium")
  } else {
    return ("Large")
  }
}

ClassifyUtilisation <- function (utilisation.percentage) {
  # Classifies the bike station depending on the proportion of bikes
  # in circulation.
  #
  # Args:
  #   utilisation.percentage: the proportion of bikes in circulation.
  #
  # Returns:
  #   Class of the station based on the utilisation percentage - three possible
  #   classes are Low, Medium and High.
  if (utilisation.percentage <= 33) {
    return ("Low")
  } else if (utilisation.percentage <= 66) {
    return ("Medium")
  } else {
    return ("High")
  } 
}

GetOpenStations <- function (df) {
  # Returns the number of stations which have their status as OPEN in a 
  # string format. This is specifically to be used for the Markdown report.
  #
  # Args:
  #   df: the data frame containing the bike info.
  #
  # Returns:
  #   A string that can be used directly in the report. In case all the
  #   stations have their status as OPEN, 'all' will be added as a prefix to
  #   the output.
  total.stations.count <- nrow(df)
  open.stations.count <- sum(df$status=="OPEN")
  result <- as.character(open.stations.count)
  if (open.stations.count == total.stations.count) {
    result <- paste('all', result)
  }
  return(result)
}

GetDistanceAndTime <- function (origins, destinations) {
  # Returns the distance and time of journey taking origins and destinations
  # as inputs. Hits Google's distance matrix API for this.
  #
  # Args:
  #   origins: latlong of origins
  #   destinations: latlong of destinations
  #
  # Returns:
  #   A vector of which the first element is the distance to the destination
  #   from the origin.
  google.distance.matrix.url <- paste0("http://maps.googleapis.com/maps/api/",
                                       "distancematrix/json?origins=",
                                       origins[1], ",", origins[2],
                                       "&destinations=",
                                       destinations[1], ",", destinations[2])
  distance.data <- GET(google.distance.matrix.url)
  distance.data.json <- content(distance.data, as="text", encoding="UTF-8")
  distance.data.df <- fromJSON(distance.data.json)
  distance <- distance.data.df$rows$elements[[1]]$distance$value
  duration <- distance.data.df$rows$elements[[1]]$duration$value
  return(c(distance, duration))
}

AddDistanceFromHere <- function (df, origin) {
  # A wrapper that calls GetDistanceAndTime().
  #
  # Args:
  #   df: the data frame containing the bike info.
  #   origin: latlong of origin.
  #
  # Returns:
  #   The inputted data frame with a new column added to it, distance. This is
  #   the distance to each station from the inputted origin latlong.
  df <- df %>% rowwise() %>%
    mutate(distance=GetDistanceAndTime(origins=origin,
                                       destinations=c(position_lat,
                                                      position_long))[1])
  return(df)
}

GetNearestStationsInfo <- function (df, origin, drop) {
  # Returns a data frame containing the info on 5 of the nearest bike stations.
  #
  # Args:
  #   df: the data frame containing the bike info.
  #   origin: latlong of origin.
  #   drop: a Boolean variable indicating whether the user wants to drop off a
  #         bike or borrow one. drop = TRUE will indicate that she is looking
  #         to return a borrowed bike.
  #
  # Returns:
  #   A data frame containing the info on 5 of the nearest bike stations
  df <- AddDistanceFromHere(df, origin)
  if (drop) {
    df <- df %>% 
      select(name, distance, available_bike_stands) %>%
      filter(available_bike_stands>0) %>%
      mutate(name_availability=paste0(
        name, " (", available_bike_stands, ")"
      ))
  } else {
    df <- df %>%
      select(name, distance, available_bikes) %>%
      filter(available_bikes>0) %>%
      mutate(name_availability=paste0(
        name, " (", available_bikes, ")"
      ))
  }
  return(data.frame(head(df, 5)))
}

PlotDistanceAvailability <- function (df, drop) {
  # Plots the bikes/bike stands available for the five nearest stations taking
  # into consideration the availability of the required resource(bike/bike
  # stand).
  #
  # Args:
  #   df: a data frame containing the info on the nearest bike stations
  #   drop: a Boolean variable indicating whether the user wants to drop off a
  #         bike or borrow one. drop = TRUE will indicate that she is looking
  #         to return a borrowed bike.
  if (drop) {
    x.label <- 'Stations (number of stands available)'
    title <- 'Stations and availability of stands'
  } else {
    x.label <- 'Stations (number of bikes available)'
    title <- 'Stations and availability of bikes'
  }
  qplot(x=reorder(name_availability, distance),
        y=distance, data=df, fill=distance, geom='blank', main=title) +
    geom_bar(stat='identity', position='dodge') +
    xlab(x.label) +
    ylab('Distance in metres') + 
    labs(fill = "Distance")
}

PlotNearestAvailableStations <- function (df, origin="Spire", drop=FALSE) {
  # Plots the bikes/bike stands available for the five nearest stations taking
  # into consideration the availability of the required resource(bike/bike
  # stand).
  #
  # Args:
  #   df: a data frame containing the info on the nearest bike stations
  #   origin: A string which will be used to find the location. dublin is added
  #           to the string to make it more specific. By default, the value is
  #           set to 'Spire'
  #   drop: a Boolean variable indicating whether the user wants to drop off a
  #         bike or borrow one. drop = TRUE will indicate that she is looking
  #         to return a borrowed bike. By default, the value is set to FALSE. 
  origin = tolower(origin)
  if (!grepl('dublin', origin)) {
    origin <- paste(origin, "dublin")
  }
  origin.coordinates <- geocode(origin)
  origin.latlong <- c(origin.coordinates[1, 2], origin.coordinates[1, 1])
  nearest.stations <- GetNearestStationsInfo(open.stations,
                                             origin.latlong,
                                             drop)
  PlotDistanceAvailability(nearest.stations, drop)
}

# Going ahead with Dublin
dublin.url <- paste0("https://api.jcdecaux.com/vls/v1/stations",
                    "?contract=Dublin",
                    "&apiKey=",
                    kJCDAPIKey)
dublin.data <- GET(dublin.url)
dublin.data.json <- content(dublin.data, as="text", encoding="UTF-8")
dublin.data.df <- fromJSON(dublin.data.json)
open.stations <- subset(dublin.data.df, status=="OPEN")
total.stands.count <- sum(open.stations$bike_stands)

position <- open.stations$position
open.stations$position <- NULL
open.stations$position_lat <- position$lat
open.stations$position_long <- position$lng

open.stations <- mutate(open.stations,
                         class=sapply(bike_stands, ClassifyStation),
                         utilisation_percentage=round(
                           available_bike_stands*100/bike_stands, 2),
                         utilisation_class=sapply(
                           utilisation_percentage, ClassifyUtilisation))
open.stations$class <- factor(open.stations$class,
                               levels=c('Small', 'Medium', 'Large'),
                               ordered=TRUE)
open.stations$utilisation_class <- factor(open.stations$utilisation_class,
                                           levels=c('Low', 'Medium', 'High'),
                                           ordered=TRUE)

utilisation.by.size <- table(open.stations$class,
                             open.stations$utilisation_class)
utilisation.by.size <- as.data.frame.matrix(utilisation.by.size)

open.stations <- group_by(open.stations, class)
class.summary <- summarise(open.stations,
                           total_bike_stands=sum(bike_stands),
                           total_available_stands=sum(available_bike_stands),
                           total_available_bikes=sum(available_bikes),
                           percentage_used=round(
                             total_available_stands*100/total_bike_stands, 2))
open.stations <- ungroup(open.stations)
class.summary <- select(class.summary, class, total_bike_stands,
                        percentage_used)
colnames(class.summary) <- c('Class', 'Total bike stands',
                             'Percentage utilisation')


Footfall.R
library(dplyr)
library(ggplot2)
library(lubridate)
library(readODS)
library(readr)
library(reshape2)
library(rvest)
library(stringr)

kColNames <- c('time', 'mon_in', 'mon_out', 'tue_in', 'tue_out', 'wed_in',
               'wed_out', 'thu_in', 'thu_out', 'fri_in', 'fri_out', 'sat_in',
               'sat_out', 'sun_in', 'sun_out')
kEntranceOne <- "O'Connell Street at Clerys"
kEntranceTwo <- "South King St"

GetEntranceData <- function (sheet, entrance.name) {
  # Classifies the time into 8 buckets of 3 hours each and returns the bucket.
  #
  # Args:
  #   sheet: the ODS sheet containing a week's data of all the entrances.
  #   entrance.name: name of the entrance of which footfall data is to be
  #   extracted.
  #
  # Returns:
  #   A data frame containing the hourly traffic data of the relevant data
  #   frame.
  search.string <- paste("Entrance Name:", entrance.name)
  index <- which(sheet$time == search.string)
  start.index <- index + 3
  end.index <- index + 26
  entrance.df <- sheet[start.index:end.index, ]
  return(entrance.df)
}

GetWeekwiseData <- function (entrance.df, direction = 'in') {
  # Returns the molten weekwise data in the direction specified.
  #
  # Args:
  #   entrance.df: the data frame containing the info of an entrance.
  #   direction: direction of the traffic, either in or out.
  #
  # Returns:
  #   A data frame containing the molten weekwise data in the direction
  #   specified.
  out.variables <- c('time', 'mon_out', 'tue_out', 'wed_out', 'thu_out',
                     'fri_out', 'sat_out', 'sun_out')
  in.variables <- c('time', 'mon_in', 'tue_in', 'wed_in', 'thu_in', 'fri_in',
                    'sat_in', 'sun_in')
  col.nums <- 1:16
  names(col.nums) <- colnames(entrance.df)
  if (direction == 'in') {
    exclude.list <- col.nums[out.variables]
  } else {
    exclude.list <- col.nums[in.variables]
  }
  weekwise.df <- entrance.df %>%
    select(-exclude.list) %>%
    melt(id.vars = 'week_num') %>%
    group_by(week_num) %>%
    mutate(value = as.integer(value))
  if (direction == 'in') {
    weekwise.df <- summarise(weekwise.df, weekly_in = sum(value))
  } else {
    weekwise.df <- summarise(weekwise.df, weekly_out = sum(value))
  }
  return(weekwise.df)
}

ReadFromODSFile <- function () {
  # Reads from the ODS file all the data pertaining to the two entrances
  # chosen and writes it to 2 CSVs. This needs to be run only for the first
  # time. CSV files can be used for the Markdown report.
  #
  # THIS FUNCTION TOOK ~9.5 MINUTES TO RUN.
  #
  # Returns:
  #   A data frame containing the molten weekwise data in the direction
  #   specified.  
  entrance.one.data <- data.frame()
  entrance.two.data <- data.frame()
  for (sheet.num in 1:52) {
    sheet <- read_ods("pedestrianfootfall2013.ods", range = "A1:O100",
                      sheet = sheet.num)
    colnames(sheet) <- kColNames
    entrance.one.week.data <- GetEntranceData(sheet, kEntranceOne)
    entrance.one.week.data <- cbind(entrance.one.week.data, week_num=sheet.num)
    entrance.one.data <- rbind(entrance.one.data, entrance.one.week.data)
    
    entrance.two.week.data <- GetEntranceData(sheet, kEntranceTwo)
    entrance.two.week.data <- cbind(entrance.two.week.data, week_num=sheet.num)
    entrance.two.data <- rbind(entrance.two.data, entrance.two.week.data)
  }
  entrance.one.data <<- entrance.one.data
  entrance.two.data <<- entrance.two.data
  write_csv(entrance.one.data, 'entrance_one_data.csv')
  write_csv(entrance.two.data, 'entrance_two_data.csv')
}

ReadFromCSVFiles <- function () {
  # Reads from the two relevant CSVs and creates 2 global variables.
  entrance.one.data <<- read_csv('entrance_one_data.csv')
  entrance.two.data <<- read_csv('entrance_two_data.csv')
}

PlotWeekWiseComparison <- function () {
  # Plots the weekwise footfall patterns over the year for both the entrances.
  e1.weekwise.in <- GetWeekwiseData(entrance.one.data)
  e2.weekwise.in <- GetWeekwiseData(entrance.two.data)
  e1.weekwise.out <- GetWeekwiseData(entrance.one.data, direction='out')
  e2.weekwise.out <- GetWeekwiseData(entrance.two.data, direction='out')
  
  e1.weekwise <- inner_join(e1.weekwise.in, e1.weekwise.out, by = 'week_num') %>%
    melt(id.vars = 'week_num')
  e2.weekwise <- inner_join(e2.weekwise.in, e2.weekwise.out, by = 'week_num') %>%
    melt(id.vars = 'week_num')
  e1.weekwise <- cbind(e1.weekwise, entrance=kEntranceOne)
  e2.weekwise <- cbind(e2.weekwise, entrance=kEntranceTwo)
  weekwise <- rbind(e1.weekwise, e2.weekwise)
  weekwise$variable = factor(weekwise$variable, labels = c('In', 'Out'),
                             levels = c('weekly_in', 'weekly_out'))
  qplot(week_num, value, data=weekwise, geom='line', colour=variable) +
    facet_grid(. ~ entrance) +
    xlab('Week') + ylab('Footfall') + labs(colour = 'Direction') +
    labs(title = 'Footfall over the weeks')
}

ExtractExchangeRate <- function(exchange.string) {
  # Extracts and returns the exchange rate as a double from the string format
  # present in the table from poundSterlingLIVE.
  #
  # Args:
  #   exchange.string: the string containing the exchange rate information
  #                    as present in the table scraped from poundSterlingLIVE.
  #
  # Returns:
  #   The exchange rate (EUR against GBP) in double format.
  exchange.string <- str_split(exchange.string, " = ", simplify = TRUE)[, 2]
  exchange.string <- substr(exchange.string, 1, nchar(exchange.string) - 4)
  exchange.rate <- round(1 / as.double(exchange.string), 3)
  return(exchange.rate)
}

PlotFootfallVsExchangeRateTrend <- function (exchange.df) {
  # Plots the trends of Euro exchange rate and footfall at entrance 1
  # (Clerys) over the 52 weeks.
  #
  # Args:
  #   exchange.df: the data frame containing the datewise exchange rate.
  e1.weekwise.in <- GetWeekwiseData(entrance.one.data)
  
  footfall.and.exchange.rates <- exchange.df %>%
    inner_join(e1.weekwise.in, by = 'week_num') %>%
    mutate(scaled_exchange_rate = round(scale(exchange_rate)[, 1], 2)) %>%
    mutate(scaled_footfall = round(scale(weekly_in)[, 1], 2)) %>%
    select(-c(exchange_rate, weekly_in)) %>%
    melt(id.vars = 'week_num')
  footfall.and.exchange.rates$variable <- factor(
    footfall.and.exchange.rates$variable,
    levels = c('scaled_exchange_rate', 'scaled_footfall'),
    labels = c('EUR value', "Footfall at Clerys (In)"))
  qplot(week_num, value, data=footfall.and.exchange.rates,
        geom='line', colour=variable) +
    xlab('Week') + ylab('Footfall and EUR value') + labs(colour = 'Measure') +
    labs(title = "A comparison of trends") +
    labs(subtitle = "Footfall at Clerys, O'Connell St. and EUR rate against GBP")
}

PlotFootfallVsExchangeRateScatter <- function (exchange.df) {
  # Plots a scatterplot of Euro exchange rate and footfall at entrance 1
  # (Clerys) over the 52 weeks.
  #
  # Args:
  #   exchange.df: the data frame containing the datewise exchange rate.
  e1.weekwise.in <- GetWeekwiseData(entrance.one.data)
  footfall.and.exchange.rates <- exchange.df %>%
    inner_join(e1.weekwise.in, by = 'week_num')
  qplot(exchange_rate, weekly_in, data=footfall.and.exchange.rates) +
    xlab('EUR value') + ylab("Footfall at Clerys") +
    labs(colour = 'Measure') +
    labs(title = "Footfall at Clerys, O'Connell St vs EUR rate against GBP")
}

dublinked.url <- paste0("https://data.dublinked.ie/dataset/8204be0a-6348-459e",
                        "-96e9-65bb75600ec3/resource/384fe47a-2f25-4f52-8fc5",
                        "-8e61899951e9/download/pedestrianfootfall2013.ods")
download.file(dublinked.url, "pedestrianfootfall2013.ods")

entrance.one.data <- data.frame()
entrance.two.data <- data.frame()

ReadFromCSVFiles()

# 4.2
# Scraping the data
exchange.url <- paste0("https://www.poundsterlinglive.com/",
                       "bank-of-england-spot/historical-spot-exchange-rates/",
                       "gbp/GBP-to-EUR-2013#charts")
exchange.data <- read_html(exchange.url)
table.nodes <- html_nodes(exchange.data, 'table')
exchange.tables <- html_table(table.nodes)
exchange.df <- exchange.tables[[length(exchange.tables)]]

colnames(exchange.df) <- tolower(colnames(exchange.df))
exchange.df <- exchange.df %>%
  mutate(date = as.POSIXct(strptime(date, format = "%a, %d %b %Y"))) %>%
  mutate(week_num = week(date)) %>%
  filter(week_num < 53) %>%
  mutate(exchange_rate = ExtractExchangeRate(rate)) %>%
  select(-c(currency, rate, date)) %>%
  group_by(week_num) %>%
  summarise(exchange_rate = round(mean(exchange_rate), 3))


Reflection.R
library(chron)
library(dplyr)
library(ggplot2)
library(lubridate)
library(readtext)
library(stringr)

kAssignmentDirectory <- paste0("~/MSc in Data Analytics/Working with Data/",
                               "Assignments/1")
kMidnight <- strptime(as.character('00:00:00'), format = "%H:%M:%S")
kThreeAm <- strptime(as.character('03:00:00'), format = "%H:%M:%S")
kSixAm <- strptime(as.character('06:00:00'), format = "%H:%M:%S")
kNineAm <- strptime(as.character('09:00:00'), format = "%H:%M:%S")
kNoon <- strptime(as.character('12:00:00'), format = "%H:%M:%S")
kThreePm <- strptime(as.character('15:00:00'), format = "%H:%M:%S")
kSixPm <- strptime(as.character('18:00:00'), format = "%H:%M:%S")
kNinePm <- strptime(as.character('21:00:00'), format = "%H:%M:%S")
kBucket1 <- 'Midnight to 3 AM'
kBucket2 <- '3 AM to 6 AM'
kBucket3 <- '6 AM to 9 AM'
kBucket4 <- '9 AM to Noon'
kBucket5 <- 'Noon to 3 PM'
kBucket6 <- '3 PM to 6 PM'
kBucket7 <- '6 PM to 9 PM'
kBucket8 <- '9 PM +'

ComputeTime <- function (command.time) {
  # Converts the timestamp into time. Uses chron package.
  #
  # Args:
  #   command.time: timestamp recorded in milliseconds; origin is 1970-01-01.
  #
  # Returns:
  #   The time in hh:mm:ss format.
  origin <- strptime("1/1/1970 00:00", "%d/%m/%Y %H:%M")
  date.time <- origin + (as.double(command.time) / 1000)
  date.time.string <- as.character(date.time)
  time.string <- substr(date.time.string, 12, 19)
  return(chron(times=time.string))
}

ClassifyTime <- function (command.time) {
  # Classifies the time into 8 buckets of 3 hours each and returns the bucket.
  #
  # Args:
  #   command.time: time represented in string format in hh:mm:ss format.
  #
  # Returns:
  #   The bucket in which command.time falls.
  command.time <- strptime(command.time, format = "%H:%M:%S")
  bucket <- kBucket1
  
  # TODO: Normal if elseif else blocks not working. Look into this.
  bucket <- ifelse(command.time < kThreeAm,
                   kBucket1,
                   ifelse(command.time < kSixAm,
                          kBucket2,
                          ifelse(command.time < kNineAm,
                                 kBucket3,
                                 ifelse(command.time < kNoon,
                                        kBucket4,
                                        ifelse(command.time < kThreePm,
                                               kBucket5,
                                               ifelse(command.time < kSixPm,
                                                      kBucket6,
                                                      ifelse(
                                                        command.time < kNinePm,
                                                             kBucket7,
                                                             kBucket8)))))))
  return(bucket)
}

PlotCommandsInBuckets <- function (commands.df) {
  # Plots the number of commands run in the console against the time buckets.
  #
  # Args:
  #   commands.df: the data frame containing the relevant commands and time of
  #                running.
  commands.df.2 <- commands.df %>%
    mutate(time_bucket = ClassifyTime(time)) %>%
    mutate(date_bucket = ifelse(date < '2017-11-20', 'Till Nov 19',
                                'After Nov 19'))
  commands.df.2$time_bucket <- factor(commands.df.2$time_bucket,
                                      levels = c(kBucket1, kBucket2, kBucket3,
                                                 kBucket4, kBucket5, kBucket6,
                                                 kBucket7, kBucket8))
  commands.df.2$date_bucket <- factor(commands.df.2$date_bucket,
                                      levels = c('Till Nov 19',
                                                 'After Nov 19'))
  
  time_bucket_analysis <- commands.df.2 %>%
    group_by(date_bucket, time_bucket) %>%
    summarise(commands.count = sum(!is.na(command)))
  
  qplot(x=time_bucket, y=commands.count, data = time_bucket_analysis,
        group=time_bucket,
        geom = 'blank', main = 'At what times am I less lazy?') +
    geom_line(aes(group = 1)) +
    # geom_bar(stat='identity', position = 'dodge', alpha = 0.4) +
    xlab("Time") + ylab('No. of R commands') +
    theme(axis.text.x = element_text(angle=90)) +
    facet_grid(. ~ date_bucket)
}

PlotCommandsVsDates <- function (commands.df) {
  # Plots the number of commands run in the console against the dates they were
  # run on.
  # 
  # Args:
  #   commands.df: the data frame containing the relevant commands and time of
  #                running.
  commands.df.2 <- commands.df %>%
    mutate(date=as.character(date)) %>%
    group_by(date) %>%
    summarise(commands.count = sum(!is.na(command)))
  
  qplot(x = date, y = commands.count, data = commands.df.2,
        group = date,
        geom = 'blank', main = 'Time management') +
    geom_line(aes(group = 1)) +
    # geom_bar(stat='identity', position = 'dodge', alpha = 0.4) +
    xlab("Time") + ylab('Number of R commands') +
    theme(axis.text.x = element_text(angle=90))
}

setwd.command <- paste0('setwd("', kAssignmentDirectory, '")')
connection <- file('history.txt', 'r')
history.txt <- readLines(con = connection)
is.assignment.directory <- FALSE

commands <- c()
times <- c()

# NUmber of lines in the imported file
initial.file.length <- length(history.txt)

# For the analysis, we need to consider only comments run while working on the
# assignment. Throughout the assignment, the working directory has been the
# same. This fact is made use of to identify which are the commands relevant
# to the analysis.
for (line.num in 1:length(history.txt)) {
  history.line <- history.txt[line.num]
  colon.index <- str_locate(history.line, ':')[1]
  time.component <- substr(history.line, 1, colon.index - 1)
  command <- substr(history.line, colon.index + 1, str_length(history.line))
  if (startsWith(command, 'setwd')) {
    if (startsWith(command, setwd.command)) {
      is.assignment.directory <- TRUE
    } else {
      if (!startsWith(command, 'setwd.command')) {
        is.assignment.directory <- FALSE
      }
    }
  }
  if (is.assignment.directory == TRUE) {
    times <- append(times, time.component)
    commands <- append(commands, command)
  }
}
close(connection)
commands.df <- data.frame(time = times, command = commands,
                          stringsAsFactors = FALSE)

commands.df <- commands.df %>%
  mutate(date = as.Date((as.double(time)) / 86400000,
                        origin = '1970-01-01')) %>%
  mutate(time = ComputeTime(time))
  

# Some of the rows can be removed
# Comments, Rmd chunk beginnings etc.
# Also, multi-line commands are represented as multiple lines in the history
# file. This also needs to be taken care of.

commands.df <- commands.df %>%
  filter(!startsWith(command, '```')) %>%
  filter(!startsWith(command, '#')) %>%
  filter(!startsWith(command, '"')) %>%
  filter(!startsWith(command, "'")) %>%
  filter(!startsWith(command, '}')) %>%
  filter(!startsWith(command, '**')) %>%
  mutate(previous_command = lag(command)) %>%
  filter(!endsWith(previous_command, ','))

final.commands.count <- nrow(commands.df)

help.used <- sum(startsWith(commands.df$command, '?'))
view.used <- sum(startsWith(commands.df$command, 'View'))
ggplot.used <- sum(startsWith(commands.df$command, 'qplot')) +
  sum(startsWith(commands.df$command, 'ggplot')) +
  sum(startsWith(commands.df$command, 'geom')) +
  sum(startsWith(commands.df$command, 'xlab')) +
  sum(startsWith(commands.df$command, 'ylab')) +
  sum(startsWith(commands.df$command, 'facet'))
dplyr.used <- sum(startsWith(commands.df$command, 'filter')) +
  sum(startsWith(commands.df$command, 'select')) +
  sum(startsWith(commands.df$command, 'mutate')) +
  sum(startsWith(commands.df$command, 'arrange')) +
  sum(startsWith(commands.df$command, 'summarise')) +
  sum(startsWith(commands.df$command, 'inner_join')) +
  sum(startsWith(commands.df$command, 'group_by')) +
  sum(startsWith(commands.df$command, 'left_join'))
require.statements <- sum(startsWith(commands.df$command, 'require'))
